pyterrier/bootstrap.py

```diff
@@ -271,14 +271,55 @@ def _index_add(self, other):
             raise ValueError("Cannot document-wise merge indices with and without positions (%r vs %r)" % (blocks_1, blocks_2))
         multiindex_cls = autoclass("org.terrier.realtime.multi.MultiIndex")
         return multiindex_cls([self, other], blocks_1, fields_1 > 0)
+    
+    def _index_corpusiter(self, direct=True):
+        def _index_corpusiter_meta(self):
+            meta_inputstream = self.getIndexStructureInputStream("meta")
+            keys = self.getMetaIndex().getKeys()
+            keys_offset = { k: offset for offset, k in enumerate(keys) }
+            while meta_inputstream.hasNext():
+                item = meta_inputstream.next()
+                yield {k : item[keys_offset[k]] for k in keys_offset}
+
+        def _index_corpusiter_direct_pretok(self):
+            meta_inputstream = self.getIndexStructureInputStream("meta")
+            keys = self.getMetaIndex().getKeys()
+            keys_offset = { k: offset for offset, k in enumerate(keys) }
+            keys_offset = {'docno' : keys_offset['docno'] }
+            direct_inputstream = self.getIndexStructureInputStream("direct")
+            lex = self.getLexicon()
+
+            while direct_inputstream.hasNext():
+                ip = direct_inputstream.getNextPostings() #this is the next() method
+
+                for skipped in range(0, direct_inputstream.getEntriesSkipped()):
+                    meta = meta_inputstream.next()
+                    yield {k : meta[keys_offset[k]] for k in keys_offset}
```

Comment:
Not super familar with the semantics of `getEntriesSkipped`. But should this include an empty `toks` to ensure that every dict from the iter has the same fields?

Reply:
it should, yes

Reply:
> Not super familar with the semantics of getEntriesSkipped.

direct_inputstream only address documents that are non-empty. So we have to address non-empty documents using getEntriesSkipped()